### PR TITLE
rng_maxbytes_period: add max-bytes/period cases for virtio-rng device

### PR DESCRIPTION
**Pull request by ybduan**
ID: 1901988

Signed-off-by: ybduan <yduan@redhat.com>

### CODE REVIEW

**Comment by ybduan** [comment]
(1/7) type_specific.io-github-autotest-qemu.rng_maxbytes_period.positive_test.scenario1: PASS (2104.77 s)
 (2/7) type_specific.io-github-autotest-qemu.rng_maxbytes_period.positive_test.scenario2: PASS (1074.67 s)
 (3/7) type_specific.io-github-autotest-qemu.rng_maxbytes_period.positive_test.scenario3: PASS (1028.14 s)
 (4/7) type_specific.io-github-autotest-qemu.rng_maxbytes_period.negative_test.scenario4: PASS (52.21 s)
 (5/7) type_specific.io-github-autotest-qemu.rng_maxbytes_period.negative_test.scenario5: PASS (4.34 s)
 (6/7) type_specific.io-github-autotest-qemu.rng_maxbytes_period.negative_test.scenario6: PASS (4.33 s)
 (7/7) type_specific.io-github-autotest-qemu.rng_maxbytes_period.negative_test.scenario7: PASS (4.33 s)

**Comment by ybduan** [comment]
Please help review this PR, thanks! @yanan-fu

**Comment by ybduan** [comment]
Please help review this PR, thanks! @luckyh

File: qemu/tests/cfg/rng_maxbytes_period.cfg [hunk 1]
```diff
@@ -0,0 +1,46 @@
...
+    check_rngd_service = "systemctl status rngd"
+    start_rngd_service = "systemctl start rngd"
+    RHEL.6:
+        driver_name = virtio
+        check_rngd_service = "service rngd status"
+        start_rngd_service = "service rngd start"
+    variants:
+        - positive_test:
+            variants:
+                - scenario1:
```
Comment (zhencliu) [inline qemu/tests/cfg/rng_maxbytes_period.cfg:17]: what about max_bytes_1k_period_1k, or something else, which can be more meaningful than scenario1/2/3...

File: qemu/tests/cfg/rng_maxbytes_period.cfg [hunk 2]
```diff
@@ -0,0 +1,46 @@
...
+                    max-bytes_virtio-rng-pci = 512
```
Comment (zhencliu) [inline qemu/tests/cfg/rng_maxbytes_period.cfg:18]: the mix use '-' and '_' could not be good, usually, we use '_' for a param definition.

File: qemu/tests/rng_maxbytes_period.py [hunk 3]
```diff
@@ -0,0 +1,85 @@
...
+
+    timeout = float(params.get("login_timeout", 360))
+    read_rng_timeout = float(params.get("read_rng_timeout", 3600))
+    read_rng_cmd = params["read_rng_cmd"]
+    max_bytes = params.get("max-bytes_virtio-rng-pci")
+    period = params.get("period_virtio-rng-pci")
+
+    if not max_bytes and not period:
+        test.error("Please specify the expected max-bytes and/or period.")
+    if not max_bytes or not period:
```
Comment (zhencliu) [inline qemu/tests/rng_maxbytes_period.py:41]: Not necessary, for line 39 makes sure at least one param is set

File: qemu/tests/rng_maxbytes_period.py [hunk 4]
```diff
@@ -0,0 +1,85 @@
...
+        if max_bytes != '0':
+            error_info = params["expected_error_info"]
+            try:
+                env_process.process(test, params, env,
+                                    env_process.preprocess_image,
+                                    env_process.preprocess_vm)
+            except virt_vm.VMCreateError as e:
```
Comment (zhencliu) [inline qemu/tests/rng_maxbytes_period.py:48]: I am afraid we cannot mix the positive test and negative test here, please consider the following situation:
what will happen if the positive test triggers an exception here? 
for the positive test, we call env_process_process(), and expect no issue, so we need to test  if any exception will be caught, and test always failed if any exception is caught. Here, if error_info not in e, the test passed...

also, we can use str(e) here

File: qemu/tests/rng_maxbytes_period.py [hunk 5]
```diff
@@ -0,0 +1,85 @@
...
+    def _is_rngd_running():
+        """
+        Check whether rngd is running
+        """
+        output = session.cmd_output(check_rngd_service)
+        if 'running' not in output:
+            return False
+        return True
+
+    timeout = float(params.get("login_timeout", 360))
```
Comment (zhencliu) [inline qemu/tests/rng_maxbytes_period.py:33]: params.get_numeric('login_timeout', 300)

File: qemu/tests/rng_maxbytes_period.py [hunk 6]
```diff
@@ -0,0 +1,85 @@
...
+                              "output: %s" % (error_info, e.output))
+            return
+
+    vm = env.get_vm(params["main_vm"])
+    vm.verify_alive()
+    session = vm.wait_for_login(timeout=timeout)
+
+    error_context.context("Read virtio-rng device to get random number",
+                          logging.info)
+    check_rngd_service = params.get("check_rngd_service")
```
Comment (zhencliu) [inline qemu/tests/rng_maxbytes_period.py:60]: check_rngd_service is always defined in global, so params['check_rngd_service'] is OK

File: qemu/tests/rng_maxbytes_period.py (hunk 4 above)
Comment (zhencliu) [reply qemu/tests/rng_maxbytes_period.py:48]: We can divide the test into two parts:
do_positive_test() 
do_negative_test()
In positive test function, do the test and check the rate; while in negative test function, do the test and check the proper error message(for ShellTimeout exception, we can still check an error msg (expected_error_info = timeout xxxx), which can keep the same check with other negative settings

just FYI.

File: qemu/tests/cfg/rng_maxbytes_period.cfg (hunk 1 above)
Comment (ybduan) [reply qemu/tests/cfg/rng_maxbytes_period.cfg]: yeah, that's acceptable. Updated, thanks!

File: qemu/tests/cfg/rng_maxbytes_period.cfg (hunk 2 above)
Comment (ybduan) [reply qemu/tests/cfg/rng_maxbytes_period.cfg:18]: This is the restriction in the avocado-vt, please check function 'add_virtio_rng' in file 'qemu_vm.py'.
The prefix 'max-bytes'  is the option name of virtio-rng-pci device in qemu command line, and 'virtio-rng-pci' is the required suffix in function 'add_virtio_rng'.
Thanks!

File: qemu/tests/rng_maxbytes_period.py (hunk 3 above)
Comment (ybduan) [reply qemu/tests/rng_maxbytes_period.py:41]: Line 39: both max_bytes and period are all not defined.
Line 41: only one parameter is defined (either max_bytes or period), negative_test will fall into this part.
Line 54: both max_bytes and period are all defined, positive_test will fall into this part.
Thanks!

Comment (zhencliu) [reply qemu/tests/rng_maxbytes_period.py:41]: Got it, so this is part of negative test, and line 68 is another part, and two 'return' for tc termination

Then please ignore my comment right below this one.

File: qemu/tests/rng_maxbytes_period.py [hunk 7]
```diff
@@ -0,0 +1,85 @@
...
+    :param test: QEMU test object
+    :param params: Dictionary with the test parameters
+    :param env: Dictionary with test environment.
+    """
+    def _is_rngd_running():
+        """
+        Check whether rngd is running
+        """
+        output = session.cmd_output(check_rngd_service)
+        if 'running' not in output:
```
Comment (zhencliu) [inline qemu/tests/rng_maxbytes_period.py:29]: return 'running' in output

is the keyword 'running' enough for the checking? Do we need to check more , e.g. Active: active (running)

File: qemu/tests/rng_maxbytes_period.py (hunk 5 above)
Comment (ybduan) [reply qemu/tests/rng_maxbytes_period.py]: Updated, thanks!

File: qemu/tests/rng_maxbytes_period.py (hunk 6 above)
Comment (ybduan) [reply qemu/tests/rng_maxbytes_period.py]: Updated, thanks!

File: qemu/tests/rng_maxbytes_period.py (hunk 7 above)
Comment (ybduan) [reply qemu/tests/rng_maxbytes_period.py:29]: Yes, it's enough. 'running' is the core keyword in output. Thanks!

File: qemu/tests/rng_maxbytes_period.py [hunk 8]
```diff
@@ -0,0 +1,84 @@
...
+        start_rngd_service = params["start_rngd_service"]
+        status, output = session.cmd_status_output(start_rngd_service)
+        if status:
+            test.error(output)
+
+    if max_bytes == '0':
+        try:
+            s, o = session.cmd_status_output(read_rng_cmd, timeout=read_rng_timeout)
+        except ShellTimeoutError:
+            return
```
Comment (zhencliu) [inline qemu/tests/rng_maxbytes_period.py:71]: 1. except ... else test.fail("Unexpected success"), this is part of negative test, read_rng_cmd should fail, we'd better to handle the situation that the read_rng_cmd succeeds unexpectedly
2. return -> pass? Then we close session in the end

File: qemu/tests/rng_maxbytes_period.py (hunk 7 above)
Comment (zhencliu) [reply qemu/tests/rng_maxbytes_period.py:29]: what about 
return 'running' in output

**Comment by zhencliu** [comment]
> (1/7) type_specific.io-github-autotest-qemu.rng_maxbytes_period.positive_test.scenario1: PASS (2104.77 s)
>>> hi @ybduan , It takes so long time, is there a way to reduce the tc exec time? 

> (2/7) type_specific.io-github-autotest-qemu.rng_maxbytes_period.positive_test.scenario2: PASS (1074.67 s)
> (3/7) type_specific.io-github-autotest-qemu.rng_maxbytes_period.positive_test.scenario3: PASS (1028.14 s)
> (4/7) type_specific.io-github-autotest-qemu.rng_maxbytes_period.negative_test.scenario4: PASS (52.21 s)
> (5/7) type_specific.io-github-autotest-qemu.rng_maxbytes_period.negative_test.scenario5: PASS (4.34 s)
> (6/7) type_specific.io-github-autotest-qemu.rng_maxbytes_period.negative_test.scenario6: PASS (4.33 s)
> (7/7) type_specific.io-github-autotest-qemu.rng_maxbytes_period.negative_test.scenario7: PASS (4.33 s)

File: qemu/tests/rng_maxbytes_period.py (hunk 7 above)
Comment (ybduan) [reply qemu/tests/rng_maxbytes_period.py]: Updated, thanks!

File: qemu/tests/rng_maxbytes_period.py (hunk 8 above)
Comment (ybduan) [reply qemu/tests/rng_maxbytes_period.py]: Updated, thanks!

**Comment by ybduan** [comment]
(01/18) type_specific.io-github-autotest-qemu.rng_maxbytes_period.positive_test.maxbytes_512_period_1k: PASS (260.04 s)
 (02/18) type_specific.io-github-autotest-qemu.rng_maxbytes_period.positive_test.maxbytes_1024_period_1k: PASS (151.65 s)
 (03/18) type_specific.io-github-autotest-qemu.rng_maxbytes_period.positive_test.maxbytes_1024k_period_1m: PASS (1037.49 s)
 (04/18) type_specific.io-github-autotest-qemu.rng_maxbytes_period.negative_test.maxbytes_0: PASS (63.26 s)
 (05/18) type_specific.io-github-autotest-qemu.rng_maxbytes_period.negative_test.maxbytes_negative: PASS (4.63 s)
 (06/18) type_specific.io-github-autotest-qemu.rng_maxbytes_period.negative_test.period_0: PASS (4.70 s)
 (07/18) type_specific.io-github-autotest-qemu.rng_maxbytes_period.negative_test.period_negative: PASS (4.64 s)
 (08/18) type_specific.io-github-autotest-qemu.rng_maxbytes_period.positive_test.maxbytes_512_period_1k: PASS (255.08 s)
 (09/18) type_specific.io-github-autotest-qemu.rng_maxbytes_period.positive_test.maxbytes_1024_period_1k: PASS (154.43 s)
 (10/18) type_specific.io-github-autotest-qemu.rng_maxbytes_period.positive_test.maxbytes_1024k_period_1m: PASS (1037.08 s)
 (11/18) type_specific.io-github-autotest-qemu.rng_maxbytes_period.negative_test.maxbytes_0: PASS (61.01 s)
 (12/18) type_specific.io-github-autotest-qemu.rng_maxbytes_period.negative_test.maxbytes_negative: PASS (4.73 s)
 (13/18) type_specific.io-github-autotest-qemu.rng_maxbytes_period.negative_test.period_0: PASS (4.70 s)
 (14/18) type_specific.io-github-autotest-qemu.rng_maxbytes_period.negative_test.period_negative: PASS (4.75 s)
 (15/18) type_specific.io-github-autotest-qemu.rng_maxbytes_period.negative_test.maxbytes_0: PASS (70.09 s)
 (16/18) type_specific.io-github-autotest-qemu.rng_maxbytes_period.negative_test.maxbytes_negative: PASS (11.90 s)
 (17/18) type_specific.io-github-autotest-qemu.rng_maxbytes_period.negative_test.period_0: PASS (11.99 s)
 (18/18) type_specific.io-github-autotest-qemu.rng_maxbytes_period.negative_test.period_negative: PASS (11.99 s)

**Comment by zhencliu** [comment]
LGTM

**Pull request merged**